File: src/main/java/com/search/CourseSchedule_207.java
package com.search;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

public class CourseSchedule_207 {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if (numCourses == 1) return true;

        List<Integer>[] graph = new List[numCourses];
        int[] ing = new int[numCourses];
        for (int[] pre: prerequisites){
            if (graph[pre[0]] == null) graph[pre[0]] = new ArrayList<>();
            graph[pre[0]].add(pre[1]);
            ing[pre[1]]++;
        }

        int n = numCourses;
        Queue<Integer> queue = new ArrayDeque<>();
        for (int i = 0; i < numCourses; i++) {
            if (ing[i] == 0) {queue.add(i); n--;}
        }

        while (!queue.isEmpty()){
            int cur = queue.poll();
            List<Integer> nexts = graph[cur];
            if (nexts == null) continue;

            for (int next: nexts){
                ing[next]--;
                if (ing[next] == 0) {
                    queue.add(next);
                    n--;
                }
            }
        }

        return n == 0;

    }
}

File: src/main/java/com/binary_tree/CourseSchedule_207.kt
package com.binary_tree

class CourseSchedule_207 {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        val graph = ArrayList<ArrayList<Int>>(numCourses)

        // 初始化 graph
        for (i in 0 until numCourses) {
            graph.add(ArrayList())
        }

        prerequisites.forEach {
            graph[it[0]].add(it[1])
        }

        val visited = IntArray(numCourses) { 0 }
        for(i in 0 until numCourses){
            if(!dfs(graph, i, visited)){
                return false
            }
        }
        return true
    }

    fun dfs(graph: List<List<Int>>, currentCourse: Int, visited: IntArray): Boolean {
        if(visited[currentCourse] == 1){
            return false
        }

        if (visited[currentCourse] == 2) {
            return true;
        }

        visited[currentCourse] = 1
        for (i in graph[currentCourse]){
            if (!dfs(graph, i, visited)){
                return false
            }
        }

        visited[currentCourse] = 2
        return true
    }
}